File: app/schemas/receipts.py
"""Pydantic schemas for receipts API."""

from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel


class ReceiptItemResponse(BaseModel):
    id: int
    product_id: Optional[int] = None
    iiko_dish_id: Optional[str] = None
    dish_name: str
    dish_code: Optional[str] = None
    dish_group: Optional[str] = None
    dish_category: Optional[str] = None
    qty: float
    price_per_unit: Optional[float] = None
    dish_sum: float
    discount_sum: float
    return_sum: float
    cost_price: Optional[float] = None
    food_cost_percent: Optional[float] = None
    margin: Optional[float] = None

    class Config:
        from_attributes = True


class ReceiptResponse(BaseModel):
    id: int
    department_id: str
    department_name: Optional[str] = None
    open_date: date
    order_num: int
    open_time: Optional[datetime] = None
    close_time: datetime
    order_type: Optional[str] = None
    table_num: Optional[str] = None
    waiter_name: Optional[str] = None
    guest_num: Optional[int] = None
    total_sum: float
    discount_sum: float
    return_sum: float
    items_count: int
    synced_at: datetime


class ReceiptDetailResponse(ReceiptResponse):
    items: List[ReceiptItemResponse] = []


class ReceiptSyncResponse(BaseModel):
    status: str
    total_receipts: int
    total_items: int
    per_domain: list
    errors: list


class ProductSalesStats(BaseModel):
    product_id: Optional[int] = None
    iiko_dish_id: Optional[str] = None
    dish_name: str
    dish_group: Optional[str] = None
    dish_category: Optional[str] = None
    total_qty: float
    total_sum: float
    total_discount: float
    total_cost: Optional[float] = None
    receipts_count: int
    avg_price: Optional[float] = None
    avg_cost: Optional[float] = None
    avg_food_cost_pct: Optional[float] = None
    margin: Optional[float] = None


class ProductSalesStatsResponse(BaseModel):
    period_start: date
    period_end: date
    department_id: Optional[str] = None
    items: List[ProductSalesStats]
    total_revenue: float
    total_cost: float
    total_margin: float
    total_items_sold: float

File: frontend/src/types/receipts.ts
export interface ReceiptItem {
  id: number
  product_id: number | null
  iiko_dish_id: string | null
  dish_name: string
  dish_code: string | null
  dish_group: string | null
  dish_category: string | null
  qty: number
  price_per_unit: number | null
  dish_sum: number        // сумма по прайсу (DishSumInt)
  paid_sum: number        // сумма к оплате со скидкой/сервисом (DishDiscountSumInt)
  discount: number        // dish_sum - paid_sum (>0 скидка, <0 наценка/сервис)
  return_sum: number
  cost_price: number | null
  food_cost_percent: number | null
  margin: number | null
}

export interface Receipt {
  id: number
  department_id: string
  department_name: string | null
  open_date: string
  order_num: number
  close_time: string
  order_type: string | null
  table_num: string | null
  waiter_name: string | null
  guest_num: number | null
  total_sum: number       // сумма по прайсу (DishSumInt)
  paid_sum: number        // сумма к оплате со скидкой/сервисом (DishDiscountSumInt)
  discount: number        // total_sum - paid_sum (>0 скидка, <0 наценка/сервис)
  return_sum: number
  items_count: number
  synced_at: string
}

export interface ReceiptDetail extends Receipt {
  items: ReceiptItem[]
}

export interface ReceiptSyncResponse {
  status: string
  total_receipts: number
  total_items: number
  per_domain: Array<{ domain: string; receipts: number; items: number }>
  errors: string[]
}

export interface ProductSalesStats {
  product_id: number | null
  iiko_dish_id: string | null
  dish_name: string
  dish_group: string | null
  dish_category: string | null
  total_qty: number
  total_sum: number
  total_discount: number
  total_cost: number | null
  receipts_count: number
  avg_price: number | null
  avg_cost: number | null
  avg_food_cost_pct: number | null
  margin: number | null
}

export interface ProductSalesStatsResponse {
  period_start: string
  period_end: string
  department_id: string | null
  items: ProductSalesStats[]
  total_revenue: number
  total_cost: number
  total_margin: number
  total_items_sold: number
}
